### PR TITLE
OCPBUGS-38118: failed to install Nutanix OCP 4.16 cluster with DHCP network

### PR DESCRIPTION
**Pull request by yanhua121**
To fix the 4.16 installer regression issue when installing a Nutanix OCP cluster with DHCP network. The issue was hit by the customer:
https://issues.redhat.com/browse/OCPBUGS-38118

The workaround fix is to add the file "/etc/hostname" with the Machine name to the master and bootstrap ignition data.

1. clusterapi: ignition interface returns secrets
    
        Updates the clusterapi ignition interface so that it returns
        all ignition secrets. Prior to this commit, the ignition interface
        returned the bootstrap ignition data, and the provision method
        turned this data into secrets. Updating the interface to return
        all secrets, gives greater flexibility to the platform to completely
        control the ignition secrets that are created for that platform.
    
        The motivation is that some platforms such as Nutanix may need
        to create per master ignition.
    
2. nutanix clusterapi: fix the regression bug OCPBUGS-38118 

Manual testing of the fix with DHCP network passed.

### CODE REVIEW

**Comment by yanhua121** [comment]
@r4f4 @jcpowermac @rvanderp3 
This 4.16 installer Nutanix DHCP bug was reported by customer. And it is a blocker issue for the customer to create a Nutanix OCP cluster using the 4.16 installer with the DHCP network. Can you help review the PR fixing this issue? I hope we can merge it soon so as to back-port it to the 4.16 release branch.

**Comment by r4f4** [comment]
> I hope we can merge it soon so as to back-port it to the 4.16 release branch.

A reminder that we just branched, so you'll need to backport this to 4.17 first.

File: pkg/infrastructure/clusterapi/clusterapi.go
Comment (yanhua121) [reply pkg/infrastructure/clusterapi/clusterapi.go:279]: It seems that the "IgnitionProvider" interface is used to customize the bootstrap ignition data only. And here for the Nutanix use case, we want to generate the ignition data for each control-plane machine separately to create the corresponding secret manifests. So I think it's simpler to handle the Nutanix special case here instead of modify the "IgnitionProvider" interface and the "IgnitionInput" struct, for they are used by other platform providers.

Comment (patrickdillon) [reply pkg/infrastructure/clusterapi/clusterapi.go:279]: The contents of ignition (bootstrap & master) are generated during asset generation--the IgnitionProvider should not change the contents of the bootstrap ignition; rather it handles how the ignition file is stored in the cloud and provided to the bootstrap node. 

So it seems these changes should go into [the master ignition asset](https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/master.go), 

possibly here:
https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/node.go#L27

If you make the changes as in this PR, you won't see the hostnames in the ignition configs when you run `openshift-install create ignition-configs`, but if you move your changes to the ignition asset, you should see your changes reflected in those files.

Comment (yanhua121) [reply pkg/infrastructure/clusterapi/clusterapi.go:279]: @patrickdillon Thanks for the comments. I will change the code based on your recommendation.

Comment (r4f4) [reply pkg/infrastructure/clusterapi/clusterapi.go]: > But it seems not work for the master ignition data, since the machine name for machine names of the master machines are different.
@yanhua121 can you explain what you mean by this?

Comment (r4f4) [reply pkg/infrastructure/clusterapi/clusterapi.go]: Oh it's because all master machines use the same ignition file but each master has its own name?

Comment (r4f4) [reply pkg/infrastructure/clusterapi/clusterapi.go]: Was this not a problem with terraform-based provisioning?

Comment (yanhua121) [reply pkg/infrastructure/clusterapi/clusterapi.go]: @r4f4 This problem is handled the same way in the terraform provisioning, see https://github.com/openshift/installer/blob/master/data/data/nutanix/cluster/main.tf#L53-L63.

Comment (patrickdillon) [reply pkg/infrastructure/clusterapi/clusterapi.go]: Oh I see, so ignition files are generated per master... I suppose we could extend the ignition interface to allow overriding master ignition. I'm willing to enable that to unblock the regression, and will take a closer look at doing that tomorrow; but I wonder if this isn't a bug in RHCOS or the MCO? I don't think the installer (and the machine actuator) should be setting the hostname like this.

**Review by patrickdillon — CHANGES_REQUESTED** [review]

File: pkg/asset/machines/clusterapi.go
Comment (patrickdillon) [reply pkg/asset/machines/clusterapi.go:69]: > so we're reading the master manifests when generating ignition

Actually my suggestion won't solve the problem, as we need would still need to generate master ignition per host

Comment (yanhua121) [reply pkg/asset/machines/clusterapi.go]: See the latest changes with incorporation of #8868 changes.

**Review by patrickdillon — COMMENTED** [review]
Ideally we would push the interface changes and the nutanix specific changes as separate commits, particularly because we would like to remove the nutanix changes from the installer as soon as we can get a solution in either the MCO or afterburn. But I won't block on that, and we can simply remove the code in a new commit.

/approve

**Review by patrickdillon — APPROVED** [review]

**Comment by yanhua121** [comment]
@r4f4 @jcpowermac 
Can you take a look and put back the "lgtm" label for the PR to merge? Thanks.

**Comment by yanhua121** [comment]
@rvanderp3 Can you take a look and put back the "lgtm" label for the PR to merge? Thanks.

**Comment by yanhua121** [comment]
@r4f4 @rvanderp3 
The PR still needs the "acknowledge-critical-fixes-only" label to merge. Can you add this label? Thanks.

**Comment by r4f4** [comment]
@yanhua121 anyone can add the label afaik

**Pull request merged**
